### PR TITLE
HBC: Make language stanza order not matter

### PR DESCRIPTION
**Pull request by vitorgalvao**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Fixes https://github.com/Homebrew/homebrew-cask/issues/42755 by checking first for exact matches and only then moving to general matches.

I previously kept it as blocks, but that made it uglier/harder to read/edit/find mistakes.

### CODE REVIEW

File: Library/Homebrew/cask/lib/hbc/dsl.rb [hunk 1]
```diff
@@ -134,9 +134,8 @@ def language_eval
       end
 
       MacOS.languages.map(&Locale.method(:parse)).each do |locale|
-        key = @language_blocks.keys.detect do |strings|
-          strings.any? { |string| locale.include?(string) }
-        end
+        key = @language_blocks.keys.detect { |strings| strings.any? { |string| locale.eql?(string) } } ||
+              @language_blocks.keys.detect { |strings| strings.any? { |string| locale.include?(string) } }
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/dsl.rb]: Maybe we should add a helper method to `Locale`, so we can add a simple unit test. Something like:

```ruby
key = locale.detect(*@language_blocks.keys)
```

Comment (vitorgalvao) [reply Library/Homebrew/cask/lib/hbc/dsl.rb]: Gave it a try, but I don’t yet understand the full workings of `Locale` to be able to do it.

I have no problem with closing this over a better solution, though.

Comment (reitermarkus) [reply Library/Homebrew/cask/lib/hbc/dsl.rb]: Basically just move this logic into a method in `Locale`.

Comment (vitorgalvao) [reply Library/Homebrew/cask/lib/hbc/dsl.rb]: That’s what I tried and failed to do. I understand the logic of what I need to do, but the code isn’t coming out right. I can’t return the correct thing and it always goes for the default language.

Comment (reitermarkus) [reply Library/Homebrew/cask/lib/hbc/dsl.rb]: Could you maybe push the code anyways, even if it isn't working?

Comment (vitorgalvao) [reply Library/Homebrew/cask/lib/hbc/dsl.rb]: Nothing like a bit of sleep to make us understand what we were doing wrong. Got it to work though I haven’t written the test for the helper method.

File: Library/Homebrew/locale.rb [hunk 2]
```diff
@@ -68,6 +68,11 @@ def eql?(other)
   end
   alias == eql?
 
+  def detect(others)
+    others.detect { |other| eql?(*other) } ||
```
Comment (reitermarkus) [inline Library/Homebrew/locale.rb]: No need for the `*` here.

File: Library/Homebrew/locale.rb [hunk 3]
```diff
@@ -68,6 +68,11 @@ def eql?(other)
   end
   alias == eql?
 
+  def detect(others)
```
Comment (reitermarkus) [inline Library/Homebrew/locale.rb]: Use `*others` here and `locale.detect(*@language_blocks.keys)` above.

Comment (vitorgalvao) [reply Library/Homebrew/locale.rb]: That was my first try, but I got `Error: '["en"]' cannot be parsed to a Locale`.

File: Library/Homebrew/locale.rb (hunk 2 above)
Comment (vitorgalvao) [reply Library/Homebrew/locale.rb]: https://github.com/Homebrew/brew/pull/4224#discussion_r191364518

File: Library/Homebrew/locale.rb (hunk 3 above)
Comment (reitermarkus) [reply Library/Homebrew/locale.rb]: Ah right, the keys are actually arrays.

File: Library/Homebrew/locale.rb (hunk 2 above)
Comment (reitermarkus) [reply Library/Homebrew/locale.rb]: `eql?` and `include?` are not implemented for multiple arguments. It should be `other.any? { |o| eql?(o) }`.

Comment (reitermarkus) [reply Library/Homebrew/locale.rb]: Maybe rename `others` to `locale_groups` and `other` to `locales`.

File: Library/Homebrew/locale.rb [hunk 4]
```diff
@@ -68,6 +68,11 @@ def eql?(other)
   end
   alias == eql?
 
+  def detect(locale_groups)
```
Comment (reitermarkus) [inline Library/Homebrew/locale.rb]: Last thing would be a test in `locale_spec.rb` for the `zh-TW`/`zh` example mentioned in https://github.com/Homebrew/homebrew-cask/issues/42755.

Comment (vitorgalvao) [reply Library/Homebrew/locale.rb]: Done.

**Pull request merged**